<a href="https://colab.research.google.com/github/nssn96/ML_iris/blob/main/ML_iris.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML Assignmet 1 - Iris dataset


*   Name : Surya Narayanan Nadhamuni Suresh
*   UTA ID : 1001877873




In [31]:
#import statements
from sklearn.datasets import load_iris
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import itertools

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
%matplotlib inline

In [32]:
#loading the iris dataset
iris_dataset=load_iris()
print(type(iris_dataset))

<class 'sklearn.utils.Bunch'>


In [33]:
#to explore the iris_dataset dictionary and understand the different available keys and thier values
print(iris_dataset.keys())
#print(iris_dataset.data)
print(iris_dataset.target)
print(iris_dataset.frame)
print(iris_dataset.target_names)
print(iris_dataset.feature_names)

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
None
['setosa' 'versicolor' 'virginica']
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


From the above code snippet we can see that we have 

*   Four input features, which are sepal length, sepal width, petal length, petal width.
*   Three target names depending on the input features. The target names are : setosa=0, versicolor=1,virginica=2



In [34]:
#storing features and target of iris dataset separately
x = pd.DataFrame(iris_dataset.data)  # x is in matrix format
y=pd.DataFrame(iris_dataset.target)   # y is in a vector format
x.columns=['sepal_length','sepal_width','petal_length','petal_width']
y.columns=['target']

In [35]:
#to check the number/size of the data matrix
print(x.shape)

(150, 4)


So, from above code we can see that we have a total of 150 samples of input features

In [36]:
#converting to dataframe
data = pd.DataFrame(data= np.c_[iris_dataset['data'], iris_dataset['target']],columns= iris_dataset['feature_names'] + ['target'])
data['species'] = pd.Categorical.from_codes(iris_dataset.target,iris_dataset.target_names)
data.columns=['sepal_length','sepal_width','petal_length','petal_width','target','species']
data

sepal_length  sepal_width  petal_length  petal_width  target    species
0             5.1          3.5           1.4          0.2     0.0     setosa
1             4.9          3.0           1.4          0.2     0.0     setosa
2             4.7          3.2           1.3          0.2     0.0     setosa
3             4.6          3.1           1.5          0.2     0.0     setosa
4             5.0          3.6           1.4          0.2     0.0     setosa
..            ...          ...           ...          ...     ...        ...
145           6.7          3.0           5.2          2.3     2.0  virginica
146           6.3          2.5           5.0          1.9     2.0  virginica
147           6.5          3.0           5.2          2.0     2.0  virginica
148           6.2          3.4           5.4          2.3     2.0  virginica
149           5.9          3.0           5.1          1.8     2.0  virginica

[150 rows x 6 columns]

In [37]:

# plt.scatter(data.sepal_length,data.sepal_width)
# plt.xlabel("Sepal length")
# plt.ylabel("Sepal Width")
# plt.show()


In [38]:
#splitting the data into train and test set
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.1, random_state=1,stratify = y)

In [39]:
# x1=x_train.petal_length
# y1 = x_train.petal_width
# num_obs = len(x1)
# noise = np.random.normal(loc=0.0, scale=5.0, size=num_obs)
# y= 10+(2*x1)+noise

# plt.figure(figsize=(20,7))
# plt.plot(x1, y, 'o')


In [40]:
#global variable to store all the weights at the end of 100 iterations of all the model , so that can be used for testing
w_all=[]

# Training

In [41]:
#class definition for Linear regression
class LinearRegression:

  global w_all

  def fit(self,data,target,steps=100,bsize=32):
    self.alpha=0.01 #learning rate
    self.m = data.shape #num of training examples,features
    self.data = np.stack((np.ones(len(data)),data),axis=1)
    #print(self.data)
    self.target=target
    self.bias=0
    self.w = np.random.uniform(-1,1,size=(2,))
    self.loss_arr=[]
    

    for i in range(steps):
      #target_pred = self.predict(self.data)
      mse_loss=0.0
      #splitting the dataset according to the batch size
      for j in range(bsize):
        value = np.random.randint(0,len(data))
        #print("value is "+str(value))
        data_new = data[value:value+1]
        data_new = np.stack((np.ones(len(data_new)),data_new),axis=1)
        target_new = target[value:value+1]

        #print(data_new)
        dw = -(2 * (data_new.T).dot(target_new-self.predict(data_new))) / self.m
        #print('I am not here')
        self.w -= (self.alpha * dw)
        mse_loss += (np.sum((target_new-self.predict(data_new)) ** 2)) / self.m
      self.loss_arr.append(mse_loss)


    # for i in range(steps):
    #   target_pred = self.predict(self.data)


    #   # calculate gradient descent
    #   dbias = -2 * np.sum(self.target-target_pred) / self.m
    #   dw = -(2 * (self.data.T).dot(self.target-target_pred)) / self.m

    #   #least square function
    #   mse_loss =  (np.sum((self.target-target_pred) ** 2)) / self.m
    #   #mse_loss = np.square(np.subtract(self.target,target_pred)).mean()
    #   #mse_loss = mean_squared_error(self.target,target_pred)
    #   self.loss_arr.append(mse_loss)

    #   #update weights
    #   self.bias-= (self.alpha * dbias)
    #   self.w -= (self.alpha * dw)

    w_all.append(self.w)
    return self.loss_arr

  def predict(self,data):
    #return data.dot(self.w) + self.bias
    return data @ self.w


In [42]:
# # 12 diff combinations of models
# model = LinearRegression()
# print(len(x1))
# l  = model.fit(x1,y1)
# print(l)
# # step_num = range(1,101)
# # plt.plot(step_num, l, label='loss vs step')

In [ ]:
# 12 diff combinations of models
model = LinearRegression()
step_num = range(1,101)
temp = ['sepal_length','sepal_width','petal_length','petal_width']
li = list(itertools.permutations(temp,2))
cnt=1
model_num=1
for i in li:
  feature1=i[0]
  feature2=i[1]
  input = x_train[feature1]
  output = x_train[feature2]
  l  = model.fit(input,output)
  plt.figure()
  plt.title('Model Number '+str(model_num))
  plt.ylabel('Loss')
  plt.xlabel('Current step')
  plt.plot(step_num, l)
  model_num+=1
  #print(l)

plt.show()


# L2 Regularization

In [45]:


# # fitting the model with data set

# # reg_sgd = SGDRegressor(loss='squared_error', max_iter=100)
# # reg_sgd.fit(X,y)
# # y_hat_pred = reg_sgd.predict(X_test)
# backup = data



# MSE_score_train = []
# MSE_score_test = []

# mod_no = range(1,13)
# for i in mod_no:
#   model = SGDRegressor(loss="squared_error", max_iter=100)
#   model.fit(x_train, y_train)
#   y_train_hat  = model.predict(x_train)
#   y_test_hat  = model.predict(x_test)
#   MSE_score_train.append(-1 * mean_squared_error(y_train,y_train_hat))
#   MSE_score_test.append(-1 * mean_squared_error(y_test,y_test_hat))
    
# plt.plot(mod_no,  MSE_score_train, label='train')
# plt.plot(mod_no,  MSE_score_test, label='test')
# plt.xlabel("Model Number")
# plt.ylabel("MSE")
# plt.legend()

# # plt.annotate('Sweet Spot', xy = (9, -.005), xytext=(9, -.015),
# #             arrowprops = dict(facecolor='black', shrink = 0.05));